### ISSUE TITLE
Cannot compile under windows 11

### ISSUE THREAD

**Issue by Soraph** [eclipse/paho.mqtt.cpp#413]
I'm trying to compile following the instructions in the readme but it fails with `Could NOT find PahoMqttC (missing: PAHO_MQTT_C_LIBRARIES)`:

```
C:\Users\Matteo\Documents\paho.mqtt.cpp>cmake -Bbuild -H. -DCMAKE_INSTALL_PREFIX=C:\mqtt\paho-cpp -DPAHO_BUILD_SAMPLES=ON -DPAHO_WITH_SSL=OFF -DCMAKE_PREFIX_PATH=C:\mqtt\paho-c
-- Building for: Visual Studio 17 2022
-- The CXX compiler identification is MSVC 19.33.31630.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at C:/Program Files/CMake/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find PahoMqttC (missing: PAHO_MQTT_C_LIBRARIES)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindPahoMqttC.cmake:33 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  src/CMakeLists.txt:26 (find_package)


-- Configuring incomplete, errors occurred!
See also "C:/Users/Matteo/Documents/paho.mqtt.cpp/build/CMakeFiles/CMakeOutput.log".
```

Checking the `CMakeList.txt` file in `src` it is possible to see that the `find_package` instruction search for what seems to be the wrong package name: https://github.com/eclipse/paho.mqtt.cpp/blob/master/src/CMakeLists.txt#L26 :

```cmake
find_package(PahoMqttC REQUIRED)
``` 

While, according to the README of the C library you should use a different package name https://github.com/eclipse/paho.mqtt.c#building-your-application-with-cmake : 

```cmake
find_package(eclipse-paho-mqtt-c REQUIRED)
```
And a test build with this change leads to a different error:

```
C:\Users\Matteo\Documents\paho.mqtt.cpp>cmake -Bbuild -H. -DCMAKE_INSTALL_PREFIX=C:\mqtt\paho-cpp -DPAHO_BUILD_SAMPLES=ON -DPAHO_WITH_SSL=OFF -DCMAKE_PREFIX_PATH=C:\mqtt\paho-c
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - no
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Configuring done
CMake Error at src/CMakeLists.txt:126 (target_link_libraries):
  Target "paho-mqttpp3-static" links to:

    PahoMqttC::PahoMqttC

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Updating the `target_link_libraries` according to the documentation of the C library (line 126):

```cmake
    ## add dependencies to the shared library
    target_link_libraries(paho-mqttpp3-static
        PRIVATE ${LIBS_SYSTEM}
        PUBLIC eclipse-paho-mqtt-c::paho-mqtt3a Threads::Threads)
```

CMake can create the build files correctly:
```
C:\Users\Matteo\Documents\paho.mqtt.cpp>cmake -Bbuild -H. -DCMAKE_INSTALL_PREFIX=C:\mqtt\paho-cpp -DPAHO_BUILD_SAMPLES=ON -DPAHO_WITH_SSL=OFF -DCMAKE_PREFIX_PATH=C:\mqtt\paho-c
-- Building for: Visual Studio 17 2022
-- The CXX compiler identification is MSVC 19.33.31630.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - no
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/Matteo/Documents/paho.mqtt.cpp/build
```

But the system cannot actually build:
```
C:\Users\Matteo\Documents\paho.mqtt.cpp>cmake --build build/ --target install
MSBuild version 17.3.1+2badb37d1 for .NET Framework
  Building Custom Rule C:/Users/Matteo/Documents/paho.mqtt.cpp/src/CMakeLists.txt
  async_client.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/async_client.h(28,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Doc
uments\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  client.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/async_client.h(28,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Doc
uments\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  connect_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/connect_options.h(27,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\
Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  create_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/create_options.h(27,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\D
ocuments\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  disconnect_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/disconnect_options.h(26,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matt
eo\Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  iclient_persistence.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/iclient_persistence.h(27,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Mat
teo\Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  message.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/message.h(28,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Document
s\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  properties.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/properties.h(28,11): fatal  error C1083: Cannot open include file: 'MQTTProperties.h': No such file or directory [C:\Users\Matteo\
Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  response_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/response_options.h(10,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo
\Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  ssl_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/ssl_options.h(30,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Docu
ments\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  string_collection.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/string_collection.h(27,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matte
o\Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  subscribe_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/subscribe_options.h(27,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matte
o\Documents\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  token.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/token.h(28,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Documents\
paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  topic.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/topic.h(27,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Documents\
paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  will_options.cpp
C:\Users\Matteo\Documents\paho.mqtt.cpp\src\mqtt/will_options.h(29,10): fatal  error C1083: Cannot open include file: 'MQTTAsync.h': No such file or directory [C:\Users\Matteo\Doc
uments\paho.mqtt.cpp\build\src\paho-cpp-objs.vcxproj]
  Generating Code...
```

What should I do?

**Comment by fpagliughi** [comment]
The `CMAKE_INSTALL_PREFIX` should point to the top-level _build_ directory for the C library.

In other words, the C++ build should be able to find the C header files in `${CMAKE_INSTALL_PREFIX}/include` and the built C library files in `${CMAKE_INSTALL_PREFIX}/lib`.

**Comment by Soraph** [comment]
So it is ok that the `CMakeList.txt` needs to be fixed?

Why should I modify `CMAKE_INSTALL_PREFIX`? According to the documentation it is used to generate the install binaries not to search for data, maybe you confused the instruction?

I tried anyway but the result did not change, I deleted the previous build directory and the build had the same output.

The contents of the directory specified by `CMAKE_PREFIX_PATH` are:
![image](https://user-images.githubusercontent.com/1741291/201113040-3a923271-975a-44ab-9881-ef9b7df979dc.png)
![image](https://user-images.githubusercontent.com/1741291/201113083-5b76dff8-8b42-4521-a580-9f35e8ddcd27.png)

**Comment by Soraph** [comment]
As of right now, I managed to compile this repository by ignoring the lib c usage described in the readme here and by using VCPKG as suggested here: https://github.com/microsoft/vcpkg/issues/1848

This, however, with the modifications to the `CMakeList.txt` that i described before along with this one:

```cmake
    ## add dependencies to the shared library
    target_link_libraries(paho-mqttpp3-static
        PRIVATE ${LIBS_SYSTEM}
        PUBLIC eclipse-paho-mqtt-c::paho-mqtt3a eclipse-paho-mqtt-c::paho-mqtt3c eclipse-paho-mqtt-c::paho-mqtt3as eclipse-paho-mqtt-c::paho-mqtt3cs Threads::Threads)
```

And by installing `paho-mqtt:x64-windows` via VCPKG changing the cmake command into:

```
cmake -Bbuild -H. -DCMAKE_INSTALL_PREFIX=C:\mqtt\paho-cpp -DPAHO_BUILD_SAMPLES=ON -DPAHO_WITH_SSL=OFF -DCMAKE_TOOLCHAIN_FILE=<vcpkg install dir>/scripts/buildsystems/vcpkg.cmake
```

**Comment by Soraph** [comment]
Ok, I rebuilt the C library with `-DPAHO_BUILD_STATIC=ON` and the build for this repo works as expected, should I make a PR with the changes (readme + cmakelist)?

**Comment by Soraph** [comment]
Now I see why I missed it, I went straight to the Windows subsection, where the instructions are more simple an there is no indication of the static option.

The modifications to the `CMakeList.txt` still needs to happend anyway, the file is looking for the wrong export and will always fail

**Comment by Soraph** [comment]
In the end I stopped trying to make it work, the source codes between this and the C library will not cooperate with the architetture selection and overall will just gift problems and wasted hours. Ended up using vcpkg.

best of luck! o/

**Comment by fpagliughi** [comment]
Let's leave this open, so that I remember to look into this before the next release. 

I will:
- Update the Windows section of the README to clarify.
- Investigate what's going on with the C exports in Windows.

**Comment by fpagliughi** [comment]
There were a number of things related to this that were fixed in the recent v1.3.x release(s), and a few more coming with the next v1.4 release.

1. The CMake files did _not_ need to be updated for the Paho C export. This library had a custom CMake "Find" module for the library that pre-dated the C exports. So that wan't part of the underlying problem. But the next version of this lib will use the C exports and get rid of its custom Find.
2. There were issues causing problems when compiling this into an application that was built with C++17, resulting in problems with static constants. This was fixed in v1.3
3. I also didn't realize that static constants needed to be exported in a Windows DLL, and are not covered by the `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`. This was also fixed in v1.3

So, this should work with the current v1.3, but a few additional updates are coming with v1.4, including some better docs.
